File: frontend/services/analytics.ts
import { api, ApiEnvelope } from "./api";

export interface HiveCounts {
	total: number;
	active: number;
	inactive: number;
	healthy: number;
	needs_attention: number;
	weak: number;
	diseased: number;
}

export interface YieldAggregate {
	harvests: number;
	total_kg: number;
	avg_kg: number;
	max_kg: number;
	min_kg: number;
}

export interface DashboardSummary {
	hives: HiveCounts;
	yield_totals: {
		all_time: YieldAggregate;
		this_month: YieldAggregate;
		prev_month: YieldAggregate;
		change_amount: number;
		change_percent: number;
	};
	recommendations: {
		open: number;
		replace: number;
		monitor: number;
	};
}

export interface YieldTrend {
	categories: string[];
	data: number[];
}

// NEW: one shared x-axis (every hive's harvest dates, chronological)
// plus one series per hive, values aligned to that shared axis with
// `null` at indices that belong to a different hive. Powers the
// multi-line History chart.
export interface HiveTrendSeries {
	hive_id: string;
	hive_name: string;
	data: (number | null)[];
}

export interface HiveYieldTrends {
	categories: string[];
	series: HiveTrendSeries[];
}

export interface HiveHealthSlice {
	label: string;
	value: number;
	color: string;
}

export interface SeasonalComparisonRow {
	year: number;
	month: number;
	total_kg: number;
	harvests: number;
}

export const analyticsService = {
	dashboardSummary: () => api.get<DashboardSummary>("/analytics/dashboard"),

	yieldTrend: (months = 12) =>
		api.get<YieldTrend>(`/analytics/yield-trend?months=${months}`),

	// NEW: per-hive multi-line trend for the History page.
	hiveYieldTrends: (months = 12) =>
		api.get<HiveYieldTrends>(`/analytics/hive-yield-trends?months=${months}`),

	hiveHealth: () => api.get<HiveHealthSlice[]>("/analytics/hive-health"),

	// NEW: per-hive this-month totals, used by the Hives list/detail
	// so "Yield (This Month)" no longer shows "—" for every hive.
	hiveMonthlyYield: () =>
		api.get<Record<string, number>>("/analytics/hive-monthly-yield"),

	seasonalComparison: (hiveId?: string) =>
		api.get<SeasonalComparisonRow[]>(
			`/analytics/seasonal-comparison${hiveId ? `?hive_id=${hiveId}` : ""}`,
		),

	reportDataset: (params?: {
		date_from?: string;
		date_to?: string;
		hive_id?: string;
	}) => {
		const qs = new URLSearchParams();
		if (params?.date_from) qs.set("date_from", params.date_from);
		if (params?.date_to) qs.set("date_to", params.date_to);
		if (params?.hive_id) qs.set("hive_id", params.hive_id);
		const query = qs.toString();
		return api.get<Record<string, unknown>>(
			`/analytics/report${query ? `?${query}` : ""}`,
		) as Promise<ApiEnvelope<Record<string, unknown>>>;
	},
};

File: server/routes/analytics.py
from flask import Blueprint, request, g

from middleware.auth_middleware import token_required, role_required
from validators.yield_validator import validate_report_filters
from services.analytics_service import AnalyticsService
from utils.responses import success, error


analytics_bp = Blueprint("analytics", __name__, url_prefix="/api/analytics")


# ── DASHBOARD SUMMARY TILES ────────────────────
@analytics_bp.route("/dashboard", methods=["GET"])
@token_required
@role_required("beekeeper")
def dashboard_summary():
    data = AnalyticsService.dashboard_summary(g.user_id)
    return success("OK", data=data, status=200)


# ── MONTHLY YIELD TREND (Line chart) ───────────
@analytics_bp.route("/yield-trend", methods=["GET"])
@token_required
@role_required("beekeeper")
def yield_trend():
    months = request.args.get("months", default=12, type=int)
    data = AnalyticsService.monthly_yield_trend(g.user_id, months=months)
    return success("OK", data=data, status=200)


# ── PER-HIVE YIELD TREND (History page, multi-line) ────
@analytics_bp.route("/hive-yield-trends", methods=["GET"])
@token_required
@role_required("beekeeper")
def hive_yield_trends():
    months = request.args.get("months", default=12, type=int)
    data = AnalyticsService.hive_yield_trends(g.user_id, months=months)
    return success("OK", data=data, status=200)


# ── PER-HIVE THIS-MONTH YIELD (Hives page list) ─
@analytics_bp.route("/hive-monthly-yield", methods=["GET"])
@token_required
@role_required("beekeeper")
def hive_monthly_yield():
    data = AnalyticsService.hive_monthly_totals(g.user_id)
    return success("OK", data=data, status=200)


# ── HIVE HEALTH DISTRIBUTION (Doughnut chart) ──
@analytics_bp.route("/hive-health", methods=["GET"])
@token_required
@role_required("beekeeper")
def hive_health():
    data = AnalyticsService.hive_health_distribution(g.user_id)
    return success("OK", data=data, status=200)


# ── FULL REPORT DATASET (Feature 5) ────────────
@analytics_bp.route("/report", methods=["GET"])
@token_required
@role_required("beekeeper")
def report_dataset():
    cleaned, field_errors = validate_report_filters(request.args)
    if field_errors:
        return error(
            "Validation failed.",
            errors=[f"{k}: {v}" for k, v in field_errors.items()],
            status=422,
        )
    try:
        data = AnalyticsService.report_dataset(
            g.user_id,
            date_from=cleaned.get("date_from"),
            date_to=cleaned.get("date_to"),
            hive_id=cleaned.get("hive_id"),
        )
    except PermissionError as e:
        return error(str(e), status=403)
    return success("OK", data=data, status=200)


# ── SEASONAL COMPARISON (Feature 6) ────────────
@analytics_bp.route("/seasonal-comparison", methods=["GET"])
@token_required
@role_required("beekeeper")
def seasonal_comparison():
    hive_id = request.args.get("hive_id")
    data = AnalyticsService.seasonal_comparison(g.user_id, hive_id=hive_id)
    return success("OK", data=data, status=200)